### PR TITLE
Streamflow RPC fixes #796

### LINKED ISSUE #796: adding waterWatch rpc and example
Making streamflow data accessible to students with an RPC and an example project.

### PR DESCRIPTION
**Pull request by hamidzr**

### CODE REVIEW

**Comment by hamidzr** [comment]
need to solve the merge conflict in this submodule.

File: package.json [hunk 1]
```diff
@@ -57,7 +57,8 @@
     "nop": "^1.0.0",
     "q": "^1.4.1",
     "ramda": "^0.20.1",
-    "request": "^2.63.0",
+    "request": "https://registry.npmjs.org/request/-/request-2.79.0.tgz",
```
Comment (brollb) [inline package.json]: why did you change this?

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    CacheManager = require('cache-manager'),
+    trace = debug('netsblox:rpc:trends:trace'),
+    Q = require('q');
+
+let baseUrl = 'https://waterservices.usgs.gov/nwis/iv/?',
+    cache = CacheManager.caching({store: 'memory', max: 1000, ttl: 36000});
+
+
+
+// let northernLat = 38;
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: Can you remove these if they are not needed?

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+//sanitize coordinate inputs
+
+//******************************
+// factor out message sending and handling of multiple users = sendNext and Stop
+// msgs: an array of msgs stored in global scope.
+let DELAY = 250;
+let msgs = [];
+
+function sendNextMsg(socket){
+    // QUESTION is there a need to check for the socket ID? no need.
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: you should check to make sure the msg is intended for the given socket. Here is a problematic scenario:

Suppose you have two sockets, `s1` and `s2` and a message queue with `a1`, `b1`, and `b2` where messages starting with `a` are sent to `s1` and `b` messages are for `s2`. Regardless of who starts sending the messages, this code will fail:

Suppose we start with `s1`:

`s1` sends `a1`
`s2` sends `b1`
`s1` sends `b2`  --- bug!

Starting with `s2` breaks right away with `a1`

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    }
+    trace(`Stopped sending messages to user with roleId ${socket.roleId}`);
+    response.send('Stopped sending messages.');
+}
+
+// notes: dont forget to return null in your stop since we are handling the response in here.
+//**********************************
+
+// factor our the bulk of the work to provide easily readable settings for users.
+// TODO rename the function
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: you should rename this function

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        // remove those with a different roleId | dont remove other's messages
+        msgs = msgs.filter(msg => {
+            return msg.dstId != socket.roleId;
+        });
+    }
+    trace(`Stopped sending messages to user with roleId ${socket.roleId}`);
+    response.send('Stopped sending messages.');
+}
+
+// notes: dont forget to return null in your stop since we are handling the response in here.
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: you can return a promise now which will be more natural anyway

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+};
+
+WaterWatchRPC.stop = function(){
+    stopSendingMsgs(this);
+    return null;
+};
+
+
+
+if(true) { // add check for the availability of api key
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: fix this

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    socket.send(msgs.shift());
+    setTimeout(sendNextMsg,DELAY,socket);
+}
+
+//checks msg ownership, needs access to socket.
+function isMsgOwner(msg){
+    return msg.dstId = socket.roleId
+}
+
+function stopSendingMsgs($this){
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: if you are going to pass the context, use `self` instead for the variable name. The dollar sign is usually referring to something with jquery and is misleading.

Also, it might be better to just pass the socket and response instead of the context

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+    socket = $this.socket;
+    response = $this.response;
+    if (msgs) {
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: will this ever be false?

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+        // remove those with a different roleId | dont remove other's messages
+        msgs = msgs.filter(msg => {
+            return msg.dstId != socket.roleId;
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: use `!==`

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        });
+    }
+    trace(`Stopped sending messages to user with roleId ${socket.roleId}`);
+    response.send('Stopped sending messages.');
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: I wouldn't do this here - I would put this on line 192

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+                theData = {
+                    siteName: item.sourceInfo.siteName,
+                    lat: item.sourceInfo.geoLocation.geogLocation.latitude,
+                    lon: item.sourceInfo.geoLocation.geogLocation.longitude,
+                    // varName: item.variable.variableName,
+                    // varDescription: item.variable.variableDescription,
+                    unit: item.variable.unit.unitCode,
+                    value: item.values[0].value[0].value
+                };
+                trace(theData);
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: make this a better log (say what it is doing) or remove it

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+            try {
+                data = JSON.parse(data);
+            } catch(e) {
+                trace('Non-JSON data...');
+                return response.send('Bad API Result: ' + data);
+            }
+
+            //clean and store the sections we want in a form of a message
+            data.value.timeSeries.forEach(item => {
+                theData = {
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: this variable is never defined - use `let`

Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: Also a better name might be a good idea, too

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+                    siteName: item.sourceInfo.siteName,
+                    lat: item.sourceInfo.geoLocation.geogLocation.latitude,
+                    lon: item.sourceInfo.geoLocation.geogLocation.longitude,
+                    // varName: item.variable.variableName,
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: use it or remove it (and the following line)

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+                msgs.push({
+                    type: 'message',
+                    msgType: msgType,
+                    dstId: socket.roleId,
+                    content: theData
+                });
+            });
+            trace('loaded ', data.value.timeSeries.length,'  items');
+            // filterout messages ( imp if having multiple people use the shared msgs var)
+            let myMsgs = msgs.filter(msg => {
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: why don't you just use `data.value.timeSeries.length` in the log? This will be unnecessary and inefficient

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+                return msg.dstId == socket.roleId && msg.msgType == msgType
+            })
+            response.send(`Sending ${myMsgs.length} messages of ${msgType}`);
+            // start sending messages - will send other user's messages too
+            // QUESTION: can you send to whatever socket.roleId you want from any source? yes can do
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: I would remove this comment - it doesn't seem helpful

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+// turn an options object into query friendly string
+function encodeQueryData(options) {
+   let ret = [];
+   for (let d in options)
+     ret.push(encodeURIComponent(d) + '=' + encodeURIComponent(options[d]));
+   return ret.join('&');
+}
+
+// used to send feedback about errors to the caller.
+function showError(err, response) {
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: this function is used in one spot - why don't you just put these two lines where this is currently called? That is, abstracting this logic into a function doesn't seem to be gaining anything

File: src/server/rpc/procedures/waterwatch/waterwatch.js [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+            let myMsgs = msgs.filter(msg => {
+                return msg.dstId == socket.roleId && msg.msgType == msgType
+            })
+            response.send(`Sending ${myMsgs.length} messages of ${msgType}`);
+            // start sending messages - will send other user's messages too
+            // QUESTION: can you send to whatever socket.roleId you want from any source? yes can do
+            sendNextMsg(socket);
+        })
+        .catch(err => {
+            // show error
```
Comment (brollb) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: this comment doesn't tell us anything

File: src/server/rpc/procedures/waterwatch/waterwatch.js (hunk 16 above)
Comment (hamidzr) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: right now it is used once but I'm positive I'll be using it later.

File: src/server/rpc/procedures/waterwatch/waterwatch.js (hunk 10 above)
Comment (hamidzr) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: Since we are using the same stop functionality in other RPCs I had this included in this function make it more reusable .

File: src/server/rpc/procedures/waterwatch/waterwatch.js (hunk 4 above)
Comment (hamidzr) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: any suggestions my american friend?

File: src/server/rpc/procedures/waterwatch/waterwatch.js (hunk 14 above)
Comment (hamidzr) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: What would be inefficient? passing more than one argument to trace?

File: src/server/rpc/procedures/waterwatch/waterwatch.js (hunk 15 above)
Comment (hamidzr) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: I want it to be on record that this is going to start sending other user's messages too.

File: src/server/rpc/procedures/waterwatch/waterwatch.js (hunk 13 above)
Comment (hamidzr) [inline src/server/rpc/procedures/waterwatch/waterwatch.js]: I want it to be available as an option here. as a reminder maybe ?

**Comment by brollb** [comment]
Can you fix the code climate issues?

**Comment by brollb** [comment]
Can you fix the clarity and bug risk issues from code climate?

**Comment by hamidzr** [comment]
I updated the streamflow to utilize the new api consumer.

**Comment by brollb** [comment]
Can you change the `westernLat`, `easternLat`, etc, to be consistent with the earthquakes rpc? `minLatitude`, etc

**Comment by brollb** [comment]
I also got this error when using the `waterTemp` rpc: 
```
  netsblox:rpc-manager:log calling /waterwatch.waterTemp("36.241301176026795","-86.86324143906249","36.14155305991051","-87.02803636093749") +0ms
  netsblox:rpc:waterwatch:trace requesting data for https://waterservices.usgs.gov/nwis/iv/?format=json&bBox=-87.0280364%2C36.1415531%2C-86.8632414%2C36.2413012&siteType=GL%2CST%2CGW%2CGW-MW%2CSB-CV%2CLA-SH%2CFA-CI%2CFA-OF%2CFA-TEP%2CAW&siteStatus=active&parameterCd=00010 +0ms
  netsblox:rpc:waterwatch:trace request is not cached, calling external endpoint +0ms
  netsblox:rpc-manager:error Uncaught exception: TypeError: Cannot convert undefined or null to object +409ms

```

**Comment by hamidzr** [comment]
the error is from a bug on apiconsumer which should be fixed in #1056

**Comment by hamidzr** [comment]
changed the arguments

**Comment by brollb** [comment]
Cool, the arguments look better now :) 

Can you resolve the merge issues? It looks like there is a change in APIConsumer which probably doesn't belong here
